### PR TITLE
Ex 02

### PR DESCRIPTION
**Pull request by mhmdrdwn**

### CODE REVIEW

File: src/mohamed_radwan_ex/ex02/bubble_sort.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        for j in range(list_length-i-1):
+            # check if current is greater than next element
+            if data_list[j] > data_list[j+1]:
+                # modify the list with last element sorted
+                data_list[j], data_list[j+1] = data_list[j+1], data_list[j]
+    # the return value can be either as a list or tuple
+    # to return tuple, convert the list back to tuple
+    # to return list to save computation cost, return data_list
+    # since it's common sense to return the same data type as input
+    # then returned data type here is tuple
```
Comment (huynhngoc) [inline src/mohamed_radwan_ex/ex02/bubble_sort.py]: What if the input data is list? :)

File: src/mohamed_radwan_ex/ex02/file_stats.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+def char_counts(input_file):
+    # Reading the file and saving it to one string & close
+    with open(input_file, 'r', encoding="utf-8") as file:
+        data_string = file.read().replace('\n', '')
+    file.close()
```
Comment (huynhngoc) [inline src/mohamed_radwan_ex/ex02/file_stats.py]: `with` statement will make it automatically call close()

File: src/mohamed_radwan_ex/ex02/file_stats.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+def char_counts(input_file):
+    # Reading the file and saving it to one string & close
+    with open(input_file, 'r', encoding="utf-8") as file:
+        data_string = file.read().replace('\n', '')
```
Comment (huynhngoc) [inline src/mohamed_radwan_ex/ex02/file_stats.py]: We may want to count whitespace in here.

File: src/mohamed_radwan_ex/ex02/message_entropy.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        # for each value calculate p/total and log(p/total)
+        h -= val / len(message) * log2(val / len(message))
+        # negative sum because entropy equation is in minus
+        # other option is negating h after finishing the loop
+    return h
+
+
+# the last string has uppercase T which is converted to
+# lowercase in the letter_freq function
+# lowercase entropy is 3.214, uppercase entropy is 3.368
```
Comment (huynhngoc) [inline src/mohamed_radwan_ex/ex02/message_entropy.py]: Nice finding!

File: src/mohamed_radwan_ex/ex02/bubble_sort.py (hunk 1 above)
Comment (mhmdrdwn) [reply src/mohamed_radwan_ex/ex02/bubble_sort.py]: If it's input data is a list, then we can can declare another variable as "output_list" and sort the "output_list" without modifying the input list.

**Pull request merged**